README.md
### psr-formats

Formats for loading and saving data using in pulsar data processing.

Supported formats:
  - DADA

### Usage

```python
from psr_formats import DADAFile

dada_file = DADAFile("path/to/dada.dump")
dada_file.load_data()

dada_file["NCHAN"]
dada_file.nchan
dada_file["NPOL"]
dada_file.npol

dada_file.data.shape
```

### Testing

```
poetry run python -m unittest
```


ChangeLog.md
### v0.1.0

- Took format files from `pfb` repo.
- Updated tests to reflect new default values in `DADAFile` methods.

### v0.1.1

- Added `timestamp_formatter` class attribute
- Added `utc_start` property to `DADAFile` class. The getter returns a
`datetime.datetime` object, and the setter can take a `str` or a
`datetime.datetime` object.
- Added corresponding test for `DADAFile` class.

### v0.1.2

- Added `DataFile` to exports in __init__.py

### v0.1.3

- Added `loaded` flag to `DataFile`

### v0.2.0

TODO:
- Need to update tests to use significantly smaller test data files so
tests can run in CI environment.
- Need to create separate `DataFile` unittest, that steals some of the
functionality the `DADAFile` test.


test/test_util.py
import os
import unittest

import numpy as np

from psr_formats import util

current_dir = os.path.dirname(os.path.abspath(__file__))
data_dir = os.path.join(current_dir, "test_data")


class TestDumpFilterCoef(unittest.TestCase):

    @classmethod
    def setUpClass(cls):
        cls.coef = np.random.rand(1025)

    def test_dump_filter_coef(self):
        coef_str = util._dump_filter_coef(self.coef)
        self.assertTrue(len(coef_str) == len(self.coef))


class TestAddFilterInfoToHeader(unittest.TestCase):

    @classmethod
    def setUpClass(cls):
        cls.ntaps = 341
        cls.coef = np.random.rand(cls.ntaps)

    def test_add_filter_info_to_header_one_level(self):
        header = {}
        filter_info = [{
            "COEFF": self.coef,
            "OVERSAMP": "8/7",
            "NTAP": len(self.coef),
            "NCHAN_PFB": "8"
        }]
        header_updated = util.add_filter_info_to_header(header, filter_info)
        self.assertTrue(header_updated["NSTAGE"] == 1)

    def test_add_filter_info_to_header_n_level(self):
        n = 3
        header = {}
        filter_info = [{
            "COEFF": self.coef,
            "OVERSAMP": "8/7",
            "NTAP": len(self.coef),
            "NCHAN_PFB": "8"
        } for i in range(n)]
        header_updated = util.add_filter_info_to_header(header, filter_info)
        self.assertTrue(header_updated["NSTAGE"] == n)


class TestDumpDadaFile(unittest.TestCase):

    @classmethod
    def setUpClass(cls):
        cls.test_file_path = os.path.join(
            data_dir, "test.dada")
        cls.test_file_path_no_coeff = os.path.join(
            data_dir, "test.no_coeff.dada")
        cls.data = np.zeros(100)
        cls.ntaps = 341
        cls.coef = np.random.rand(cls.ntaps)

    @classmethod
    def tearDownClass(cls):
        pass

    def test_dump_dada_file_no_filter_coef(self):
        header = {"HDR_SIZE": 4096}
        util.dump_dada_file(
            self.test_file_path_no_coeff,
            header,
            self.data
        )

    def test_dump_dada_file(self):
        filter_info = [{
            "COEFF": self.coef,
            "OVERSAMP": "8/7",
            "NTAP": len(self.coef),
            "NCHAN_PFB": "8"
        }]
        header = {"HDR_SIZE": 4096}
        header = util.add_filter_info_to_header(
            header, filter_info)
        util.dump_dada_file(
            self.test_file_path,
            header,
            self.data
        )


class TestLoadDadaFile(unittest.TestCase):

    @classmethod
    def setUpClass(cls):
        cls.test_file_path_big_header = os.path.join(
            data_dir, "test_load_dada_file.big_header.dump")

        cls.test_file_path_std_header = os.path.join(
            data_dir, "test_load_dada_file.std_header.dump")

    def test_load_dada_file_big_header(self):
        header, data = util.load_dada_file(
            self.test_file_path_big_header,
        )
        self.assertTrue(header["HDR_SIZE"] == "8192")
        self.assertTrue(header["NPOL"] == "2")
        self.assertTrue(header["NCHAN"] == "1")

    def test_load_dada_file_std_header(self):
        header, data = util.load_dada_file(
            self.test_file_path_std_header,
        )
        self.assertTrue(header["HDR_SIZE"] == "4096")
        self.assertTrue(header["NPOL"] == "1")
        self.assertTrue(header["NCHAN"] == "1")


if __name__ == '__main__':
    unittest.main()


test/test_imports.py
import unittest


class TestImports(unittest.TestCase):

    def test_imports(self):
        from psr_formats import DADAFile, DataFile
        self.assertTrue(DADAFile.__name__ == "DADAFile")
        self.assertTrue(DataFile.__name__ == "DataFile")


if __name__ == "__main__":
    unittest.main()


test/test_dada_file.py
import unittest
import os
import logging
import datetime

import numpy as np

from psr_formats.dada_file import DADAFile

current_dir = os.path.dirname(os.path.abspath(__file__))
test_data_dir = os.path.join(current_dir, "test_data")


class TestDADAFile(unittest.TestCase):

    def setUp(self):
        test_dada_file_paths = [os.path.join(
            test_data_dir,
            "py_channelized.impulse.noise_0.0.nseries_3.ndim_2.os.dump"
        ), os.path.join(
            test_data_dir,
            "pre_Fold.vanilla.pulsar.10.dump"
        )]
        self.dada_files = [DADAFile(f) for f in test_dada_file_paths]

    def test_getitem(self):
        self.dada_files[0]._load_data_from_file()
        self.assertTrue(self.dada_files[0]["NCHAN"] == "8")
        self.assertTrue(self.dada_files[0]["NPOL"] == "2")
        self.assertTrue(self.dada_files[0]["NDIM"] == "2")
        self.assertTrue(self.dada_files[0]["NBIT"] == "32")

        self.dada_files[1]._load_data_from_file()
        self.assertTrue(self.dada_files[1]["NCHAN"] == "1")
        self.assertTrue(self.dada_files[1]["NPOL"] == "1")
        self.assertTrue(self.dada_files[1]["NDIM"] == "4")
        self.assertTrue(self.dada_files[1]["NBIT"] == "32")

    def test_setitem(self):
        self.dada_files[0]._load_data_from_file()
        self.dada_files[0]["NCHAN"] = "10"
        self.assertTrue(self.dada_files[0]["NCHAN"] == "10")

    def test_contains(self):
        self.dada_files[0]._load_data_from_file()
        self.assertTrue("NCHAN" in self.dada_files[0])

    def test_load_data_from_file(self):
        self.dada_files[0]._load_data_from_file()
        self.assertIsInstance(self.dada_files[0]._data, np.ndarray)
        self.assertTrue("NCHAN" in self.dada_files[0]._header)

    def test_shape_data(self):
        self.dada_files[0]._load_data_from_file()
        data_expected = self.dada_files[0]._shape_data(self.dada_files[0].data)
        self.assertTrue(
            all([i == j for i, j in zip(data_expected.shape, (443961, 8, 2))]))

        ndim, nchan, npol = [int(self.dada_files[0][item])
                             for item in ["NDIM", "NCHAN", "NPOL"]]

        data_flat = self.dada_files[0].data

        data_shaped_ = data_flat.reshape((-1, nchan, npol*ndim))
        data_shaped = np.zeros(
            (data_shaped_.shape[0], nchan, npol),
            dtype=self.dada_files[0]["COMPLEX_DTYPE"])

        data_shaped[:, :, 0] = data_shaped_[:, :, 0] + 1j*data_shaped_[:, :, 1]
        data_shaped[:, :, 1] = data_shaped_[:, :, 2] + 1j*data_shaped_[:, :, 3]

        self.assertTrue(np.allclose(data_expected, data_shaped))

    def test_load_data(self):
        self.dada_files[0].load_data()

    def test_dump_data(self):
        self.dada_files[0].load_data()
        new_file_path = self.dada_files[0].dump_data(overwrite=False)
        self.assertFalse(new_file_path == self.dada_files[0].file_path)
        os.remove(new_file_path)

    def test_get_utc_start(self):
        self.dada_files[0].load_data()
        d = self.dada_files[0].utc_start
        expected_str = "2019-02-06 03:41:41"
        self.assertTrue(str(d) == expected_str)

    def test_set_utc_start(self):
        self.dada_files[0].load_data()
        new_utc_start = datetime.datetime.utcnow()
        self.dada_files[0].utc_start = new_utc_start

        self.dada_files[0].utc_start = new_utc_start.strftime(
            DADAFile.timestamp_formatter)

        with self.assertRaises(ValueError):
            self.dada_files[0].utc_start = "foo"


if __name__ == "__main__":
    logging.basicConfig(level=logging.DEBUG)
    unittest.main()


psr_formats/dada_file.py
# dada_file.py
import logging
import os
import datetime

import numpy as np

from .data_file import DataFile
from .util import (
    load_dada_file,
    dump_dada_file
)

module_logger = logging.getLogger(__name__)


class DADAFile(DataFile):

    default_header = {
        "HDR_VERSION": "1.0",
        "HDR_SIZE": "4096",
        "TELESCOPE": "PKS",
        "PRIMARY": "dspsr",
        "UTC_START": "2007-05-18-15:55:58",
        "SOURCE": "J1644-4559",
        "RA": "16:44:49.28",
        "DEC": "-45:59:09.5",
        "FREQ": "1405.000000",
        "BW": "40",
        "TSAMP": "0.0125",
        "NBIT": "32",
        "NDIM": "2",
        "NPOL": "2",
        "NCHAN": "1",
        "MODE": "PSR",
        "OBS_OFFSET": "0",
        "INSTRUMENT": "dspsr",
        "DSB": "0",
        "PFB_DC_CHAN": "1"
    }

    timestamp_formatter = "%Y-%m-%d-%H:%M:%S"

    def __init__(self, file_path: str):
        super(DADAFile, self).__init__(file_path)
        self._header = self.default_header.copy()
        self.logger = module_logger.getChild("DADAFile")

    def _load_data_from_file(self) -> None:

        self._header, self._data = load_dada_file(self.file_path)

    def _shape_data(self, data: np.ndarray) -> None:

        if self._header is None:
            raise RuntimeError(("DADAFile._shape_data: Need to load "
                                "data from file before calling _shape_data"))

        ndim, nchan, npol = [int(self[item])
                             for item in ["NDIM", "NCHAN", "NPOL"]]

        data = data.reshape((-1, nchan, npol, ndim))
        if ndim == 2:  # means we're dealing with complex data
            data = data[:, :, :, 0] + 1j*data[:, :, :, 1]

        return data

    @property
    def utc_start(self) -> datetime.datetime:
        return datetime.datetime.strptime(
            self._header["UTC_START"], self.timestamp_formatter)

    @utc_start.setter
    def utc_start(self, new_utc_start):
        """
        Can either pass a str or a datetime object.
        If str doesn't fit `timestamp_formatter` class attribute,
        then raises ValueError
        """
        if hasattr(new_utc_start, "strftime"):  # passing datetime object
            self._header["UTC_START"] = new_utc_start.strftime(
                self.timestamp_formatter)
        else:  # passing str object
            datetime.datetime.strptime(
                new_utc_start, self.timestamp_formatter)
            self._header["UTC_START"] = new_utc_start

    def load_data(self):
        self._loaded = True
        self._load_data_from_file()
        self._data = self._shape_data(self._data).copy()
        return self

    def dump_data(self, overwrite: bool = True) -> str:

        new_file_path = self.file_path
        if not overwrite:
            exists = os.path.exists(new_file_path)
            temp_file_path = new_file_path
            # temp_file_path = f"{new_file_path}.{i}"
            i = 0
            while exists:
                temp_file_path = f"{new_file_path}.{i}"
                exists = os.path.exists(temp_file_path)
                i += 1
            new_file_path = temp_file_path

        if self.ndim == 1:
            data = self.data.flatten()
        else:
            data = np.zeros((self.ndat, self.nchan, self.ndim*self.npol),
                            dtype=np.float32)
            for pol in range(self.npol):
                data[:, :, pol*2] = self.data[:, :, pol].real
                data[:, :, pol*2 + 1] = self.data[:, :, pol].imag
        self.logger.debug(f"dump_data: new file path: {new_file_path}")
        dump_dada_file(new_file_path, self.header, data)
        return new_file_path


psr_formats/data_file.py
# data_file.py
import logging
import typing

import numpy as np


module_logger = logging.getLogger(__name__)

__all__ = [
    "DataFile",
]


class DataFile:

    def __init__(self, file_path: str):
        self.logger = module_logger.getChild("DataFile")
        self._file_path = file_path
        self._header = None
        self._data = None
        self._loaded = False

    @property
    def file_path(self) -> str:
        return self._file_path

    @file_path.setter
    def file_path(self, file_path: str) -> None:
        self._file_path = file_path
        self._data = None
        self._header = None

    @property
    def header(self) -> dict:
        return self._header

    @header.setter
    def header(self, new_header: dict) -> None:
        self._header = new_header.copy()

    @property
    def data(self) -> np.ndarray:
        return self._data

    @data.setter
    def data(self, new_data: np.ndarray):
        # if new_data.ndim != 1 and new_data.ndim != 4:
        #     raise RuntimeError(f"Ambiguous data shape: {new_data.ndim}")
        iscomplex = np.iscomplexobj(new_data)
        ndim = 2 if iscomplex else 1
        ndat = new_data.shape[0]

        if new_data.ndim == 1:
            nchan, npol = 1, 1
            self._data = np.zeros((ndat, nchan, npol), dtype=new_data.dtype)
            self._data[:, 0, 0] = new_data
        else:
            nchan, npol = new_data.shape[1], new_data.shape[2]
            self._data = new_data

        self["NDAT"] = str(ndat)
        self["NCHAN"] = str(nchan)
        self["NPOL"] = str(npol)
        self["NDIM"] = str(ndim)

    @property
    def loaded(self):
        return self._loaded

    @property
    def nchan(self) -> int:
        return int(self["NCHAN"])

    @property
    def ndim(self) -> int:
        return int(self["NDIM"])

    @property
    def npol(self) -> int:
        return int(self["NPOL"])

    @property
    def ndat(self) -> int:
        if self.data is not None:
            if self.data.ndim > 1:
                return self.data.shape[0]

    def __getitem__(self, item: str) -> typing.Any:
        if self._header is not None:
            if item in self._header:
                return self._header[item]

    def __setitem__(self, item: str, val: typing.Any) -> None:
        if self._header is not None:
            if item in self._header:
                self._header[item] = val

    def __contains__(self, item: str) -> bool:
        if self._header is not None:
            if item in self._header:
                return True
            else:
                return False
        else:
            raise RuntimeError(("DataFile.__contains__: Need to load "
                                "data from file before calling __contains__"))

    def load_data(self):
        raise NotImplementedError()

    def dump_data(self, overwrite=False):
        raise NotImplementedError()


psr_formats/util.py
import logging
import typing

import numpy as np

module_logger = logging.getLogger(__name__)

__all__ = [
    "load_dada_file",
    "dump_dada_file",
    "add_filter_info_to_header",
]

_float_dtype_map = {
    '32': np.float32,
    '64': np.float64
}

_complex_dtype_map = {
    '32': np.complex64,
    '64': np.complex128
}

_exclude_header_keys = ["COMPLEX_DTYPE", "FLOAT_DTYPE"]


def _process_header(header_arr: np.ndarray) -> dict:
    header_str = "".join([c.decode("UTF-8") for c in header_arr.tolist()])
    lines = header_str.split("\n")
    header = {}
    for line in lines:
        if line.startswith("#") or not line:
            continue
        else:
            key, val = line.split()[:2]
            header[key] = val
    return header


def load_dada_file(file_path: str) -> typing.List:
    header_size = 4096  # smallest header size supported by DADA format
    header_read = False
    with open(file_path, "rb") as file:
        buffer = file.read()

    while not header_read:
        header = np.frombuffer(
            buffer, dtype='c', count=header_size
        )
        header = _process_header(header)
        if "HDR_SIZE" in header:
            new_header_size = int(header["HDR_SIZE"])
            if new_header_size == header_size:
                header_read = True
            else:
                header_size = new_header_size
        else:
            header_size *= 2

    float_dtype = _float_dtype_map[str(header["NBIT"])]
    complex_dtype = _complex_dtype_map[str(header["NBIT"])]
    header["FLOAT_DTYPE"] = float_dtype
    header["COMPLEX_DTYPE"] = complex_dtype
    data = np.frombuffer(
        buffer, dtype=float_dtype, offset=header_size
    )
    return [header, data]


def add_filter_info_to_header(
    header: dict,
    filter_info: typing.List[dict]
) -> dict:
    nstage = len(filter_info)
    header["NSTAGE"] = nstage
    for i in range(nstage):
        filter_coef = filter_info[i]["COEFF"]
        filter_coef_str = ",".join(
            _dump_filter_coef(filter_coef))

        header[f"OVERSAMP_{i}"] = filter_info[i]["OVERSAMP"]
        header[f"NTAP_{i}"] = len(filter_coef)
        header[f"COEFF_{i}"] = filter_coef_str
        header[f"NCHAN_PFB_{i}"] = filter_info[i]["NCHAN_PFB"]
    return header


def dump_dada_file(file_path: str,
                   header: dict,
                   data: np.ndarray) -> None:
    module_logger.debug(f"dump_dada_file file_path: {file_path}")
    module_logger.debug(f"dump_dada_file header: {header}")

    def header_to_str(header: dict) -> str:
        header_str = "\n".join(
            [f"{key} {header[key]}" for key in header
             if key not in _exclude_header_keys]) + "\n"
        return header_str

    header_size = int(header["HDR_SIZE"])
    header_str = header_to_str(header)
    header_len = len(header_str)
    while header_size < header_len:
        header_size *= 2
        header["HDR_SIZE"] = header_size
        header_str = header_to_str(header)
        header_len = len(header_str)

    header_bytes = str.encode(header_str)
    remaining_bytes = header_size - len(header_bytes)
    module_logger.debug(
        f"dump_dada_file len(header_bytes): {len(header_bytes)}")
    module_logger.debug(
        f"dump_dada_file remaining_bytes: {remaining_bytes}")
    header_bytes += str.encode(
        "".join(["\0" for i in range(remaining_bytes)]))

    assert len(header_bytes) == header_size, \
        f"Number of bytes in header must be equal to {header_size}"

    with open(file_path, "wb") as output_file:
        output_file.write(header_bytes)
        output_file.write(data.flatten().tobytes())


def _dump_filter_coef(filter_coef: np.ndarray) -> typing.List[str]:
    """
    Given some filter coefficients, dump them to ascii format.

    Returns:
        list: a list of strings
    """
    filter_coef_as_ascii = ["{:.6E}".format(n) for n in filter_coef]
    return filter_coef_as_ascii


# def _add_fir_data_to_existing_file(
#     file_path: str,
#     fir_file_path: str,
#     os_factor: str,
#     channels: int,
#     overwrite: bool = False
# ) -> None:
#     _, coeff = load_matlab_filter_coef(fir_file_path)
#
#     fir_info = [{
#         "COEFF": coeff,
#         "NTAPS": len(coeff),
#         "OVERSAMP": str(os_factor),
#         "NCHAN_PFB": channels
#     }]
#
#     header, data = load_dada_file(file_path)
#     header = add_filter_info_to_header(header, fir_info)
#     output_file_path = file_path
#     counter = 0
#     if not overwrite:
#         output_file_path = f"{output_file_path}.{counter}"
#         while os.path.exists(output_file_path):
#             counter += 1
#             output_file_path_split = output_file_path.split(".")
#             output_file_path_split[-1] = str(counter)
#             output_file_path = ".".join(output_file_path_split)
#
#     dump_dada_file(output_file_path, header, data)
#
#
# def create_parser():
#
#     # current_dir = os.path.dirname(os.path.abspath(__file__))
#
#     # config_dir = os.getenv("PFB_CONFIG_DIR",
#     #                        os.path.join(current_dir, "config"))
#     # data_dir = os.getenv("PFB_DATA_DIR",
#     #                      os.path.join(current_dir, "data"))
#
#     parser = argparse.ArgumentParser(
#         description="add FIR filter info to existing DADA file")
#
#     parser.add_argument("-i", "--input-file",
#                         dest="input_file_path",
#                         required=True)
#
#     parser.add_argument("-f", "--fir-file",
#                         dest="fir_file_path",
#                         required=True)
#
#     parser.add_argument("-c", "--channels",
#                         dest="channels", default=8, type=int)
#
#     parser.add_argument("-os", "--oversampling_factor",
#                         dest="oversampling_factor", default="1/1", type=str)
#
#     parser.add_argument("-ow", "--overwrite",
#                         dest="overwrite", action="store_true")
#
#     return parser
#
#
# if __name__ == "__main__":
#     parsed = create_parser().parse_args()
#     # log_level = logging.INFO
#     # if parsed.verbose:
#     #     log_level = logging.DEBUG
#     #
#     # logging.basicConfig(level=log_level)
#     # logging.getLogger("matplotlib").setLevel(logging.ERROR)
#
#     _add_fir_data_to_existing_file(
#         parsed.input_file_path,
#         parsed.fir_file_path,
#         parsed.oversampling_factor,
#         parsed.channels,
#         parsed.overwrite
#     )
